File: client/src/app/services/models/budget-subtype-dto.ts
import { BudgetTypeDto } from '../models/budget-type-dto';

/**
 * DTO representing a budget subtype, containing details about the subtype and its associated budget type.
 */
export type BudgetSubtypeDto = {

/**
 * Unique correlation ID to trace and associate requests/responses.
 */
'correlationId': string;

/**
 * Unique identifier for the entity, typically assigned upon creation or retrieval.
 */
'id': string;

/**
 * Version number for optimistic locking and concurrent updates management.
 */
'version': number;

/**
 * Name of the budget subtype.
 */
'name': string;

/**
 * Total amount spent for this budget subtype.
 */
'availableFunds'?: number | null;

/**
 * Description of the budget subtype.
 */
'description': string;

/**
 * UUID of the associated budget type.
 */
'budgetTypeId'?: string | null;
'budgetType'?: BudgetTypeDto | null;
};

File: server/portalMicroservice/src/main/java/com/portalMicroservice/dto/budget/BudgetSubtypeDTO.java
package com.portalMicroservice.dto.budget;

import com.fasterxml.jackson.annotation.JsonInclude;
import io.swagger.v3.oas.annotations.media.Schema;
import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.NotNull;
import lombok.Getter;
import lombok.Setter;

import java.util.UUID;

@Getter
@Setter
@JsonInclude(JsonInclude.Include.NON_NULL)
@Schema(description = "DTO representing a budget subtype, containing details about the subtype and its associated budget type.")
public class BudgetSubtypeDTO extends AbstractDTO {

    @NotNull
    @NotEmpty
    @Schema(description = "Name of the budget subtype.",
            example = "Marketing",
            required = true)
    private String name;

    @Schema(description = "Total amount spent for this budget subtype.",
            example = "1200.50",
            nullable = true)
    private Double availableFunds;

    @NotNull
    @NotEmpty
    @Schema(description = "Description of the budget subtype.",
            example = "Budget allocated for marketing activities.",
            required = true)
    private String description;

    @Schema(description = "UUID of the associated budget type.",
            example = "123e4567-e89b-12d3-a456-426614174000",
            nullable = true)
    private UUID budgetTypeId;

    @Schema(description = "Associated BudgetTypeDTO for this budget subtype. It provides additional details about the budget type.",
            nullable = true)
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private BudgetTypeDTO budgetType;
}
